File: dynamic_programming/도둑질.py
def solution(money):
    answer = 0
    n = len(money)
    dp_1, dp_2 = [0 for _ in range(n-1)], [0 for _ in range(n)]
    dp_1[0] = money[0]
    dp_1[1] = max(money[0], money[1])
    for i in range(2, n-1):
        dp_1[i] = max(dp_1[i-1], dp_1[i-2]+money[i])
    dp_2[1] = money[1]
    for i in range(2, n):
        dp_2[i] = max(dp_2[i-1], dp_2[i-2]+money[i])
    answer = max(dp_1[n-2], dp_2[n-1])
    return answer

File: dynamic_programming/도둑질.java
class Solution {
    public int solution(int[] money) {
        int answer = 0;
        int n = money.length;
        
        int[] dp1 = new int[n];
        int[] dp2 = new int[n];
        // 첫번째 집을 터는 대신 마지막 집을 털 수 없음
        dp1[0] = money[0];
        dp1[1] = money[0];
        // 첫번째 집을 털지 않는 대신 마지막 집을 털 수 있음
        dp2[0] = 0;
        dp2[1] = money[1];
        for (int i = 2; i < n - 1; i++) {
            dp1[i] = Math.max(money[i] + dp1[i - 2], dp1[i - 1]);
            dp2[i] = Math.max(money[i] + dp2[i - 2], dp2[i - 1]);
        }
        dp1[n - 1] = Math.max(dp1[n - 2], dp1[n - 3]);
        dp2[n - 1] = Math.max(money[n - 1] + dp2[n - 3], dp2[n - 2]);
        
        return Math.max(dp1[n - 1], dp2[n - 1]);
    }
}
